### ISSUE TITLE
BatchNormalization

### ISSUE THREAD

**Issue by prashant45** [google-coral/edgetpu#212]
Hi,

I have a model with BatchNormalization layer. 
Model - 
[model-quant.tflite.tar.gz](https://github.com/google-coral/edgetpu/files/5151097/model-quant.tflite.tar.gz)

I get the following error when compiling for edgetpu - 

```
Edge TPU Compiler version 14.1.317412892
ERROR: :334 bias->type != kTfLiteInt32 (INT8 != INT32)
ERROR: Node number 2 (CONV_2D) failed to prepare.

ERROR: :334 bias->type != kTfLiteInt32 (INT8 != INT32)
ERROR: Node number 2 (CONV_2D) failed to prepare.

Compilation failed: Internal error

Internal compiler error. Aborting!
```

Any possible way to have a workaorund for this issue? 
The layer is simply a Mul and Add opp in tflite.

**Comment by Namburger** [comment]
Humn, I inspected the model and it looks okay, anyways you can show your quantization process?

**Comment by prashant45** [comment]
My quantization,

```
converter = tf.lite.TFLiteConverter.from_keras_model(model)
input_shape = model.input.shape
inference_dtype = tf.uint8

converter.optimizations = [tf.lite.Optimize.DEFAULT]
np.random.seed(20)

def representative_dataset_gen():
    for _ in range(100):
        # Get sample input data as a numpy array in a method of your choosing.
        yield [np.array(np.random.randn(*input_shape), dtype=np.float32)]

converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.inference_input_type = inference_dtype
converter.inference_output_type = inference_dtype

converter.representative_dataset = representative_dataset_gen
```

**Comment by Namburger** [comment]
@prashant45 oh, may I ask what tensorflow version you're using?

For 2.3 and above, I believe you'll need to set the batchsize or else it'll will not be static. It is not ideal since the visualizer like netron doesn't shows that the inputs in static.

```
model.input.set_shape((1,) + model.input.shape[1:])
converter = tf.lite.TFLiteConverter.from_keras_model(model)
```
Although the error still seems a different than what I expected. It is a possible bug

**Comment by prashant45** [comment]
@Namburger this problem was solved when I use a smaller input size. For example, `[1, 4000, 1]`.

However, I incorporated some custom normalization layers for a new model. This is is not compiling. 

[model-gln-quant.tflite.tar.gz](https://github.com/google-coral/edgetpu/files/5268476/model-gln-quant.tflite.tar.gz)

**Comment by Namburger** [comment]
@prashant45 did you added the line to make the batch size static as I suggested?

**Comment by prashant45** [comment]
@Namburger no, I just used a smaller input size

**Comment by prashant45** [comment]
> @prashant45 did you added the line to make the batch size static as I suggested?

@Namburger no, I just used a smaller input-size

**Comment by Namburger** [comment]
@prashant45 lol can you try it? that was my suggestion.

**Comment by prashant45** [comment]
> @prashant45 lol can you try it? that was my suggestion.

I tired it and it did not work. Otherwise I would had mentioned it. :P

**Comment by Namburger** [comment]
I see, can you possibly attach the code to reproduce your model? it is rather weird that I've not seen this issue before

**Comment by prashant45** [comment]
> I see, can you possibly attach the code to reproduce your model? it is rather weird that I've not seen this issue before

```
import os
os.environ["CUDA_VISIBLE_DEVICES"] = "-1"
import argparse
import numpy as np
import tensorflow as tf

N = 256
L = 20
B = 256
H = 512
P = 3
X = 2
R = 4
C = 2


def model(batch_dim, input_shape):

    inputs = tf.keras.Input(shape=input_shape, batch_size=batch_dim, name="input", dtype=tf.float32)

    # Encoder
    encoder = tf.keras.layers.Conv1D(filters=N, kernel_size=L, strides=L // 2, use_bias=False, activation="relu", name="encoder_conv")(inputs)  # [B, sequence_len, channel_size]

    # Separator (Temporal ConvNet)
    layer_norm = tf.keras.layers.BatchNormalization(batch_size=batch_dim, name="layer_norm")(encoder)

    bottleneck_conv = tf.keras.layers.Conv1D(filters=B, kernel_size=1, use_bias=False, name="bottleneck_conv")(layer_norm)
    next_out = bottleneck_conv

    # Temporal block
    for r in range(R):
        for x in range(X):

            now_block = "block_{}_{}_".format(r, x)
            block_input = next_out

            next_out = tf.keras.layers.Conv1D(filters=H, kernel_size=1, use_bias=False, activation="relu", name=now_block + "1x1_conv")(block_input)
            next_out = tf.keras.layers.BatchNormalization(name=now_block + "norm_one")(next_out)
            next_out = tf.keras.layers.SeparableConv1D(filters=B, kernel_size=P, strides=1, dilation_rate=2 ** x, activation="relu", padding='same', name=now_block + "separable_conv")(next_out)
            next_out = tf.keras.layers.BatchNormalization(name=now_block + "norm_two")(next_out)

            next_out = tf.keras.layers.Add(name=now_block + "residual_add")([block_input, next_out])

    score = tf.keras.layers.Conv1D(C*N, kernel_size=1, use_bias=False, name="mask_1x1")(next_out)
    score = tf.keras.layers.Reshape(target_shape=(C, -1, N), name="score_reshape")(score)  # (C, K, N) -> K can be dynamic because of dynamic input_size
    est_mask = tf.keras.layers.Activation("relu", name="est_mask")(score)

    # Decoder
    mixture_w = tf.expand_dims(encoder, axis=1)
    source_w = tf.keras.layers.Multiply(name="source_weights")([mixture_w, est_mask])
    est_sources = tf.keras.layers.Dense(L, use_bias=False, name="basis_signals")(source_w)

    return tf.keras.Model(inputs=inputs, outputs=est_sources, name="Conv-TaskNet")


if __name__ == "__main__":

    parser = argparse.ArgumentParser()

    parser.add_argument('--save-dir', type=str,
                        help='dir to save the model')

    parser.add_argument('--model-name', type=str,
                        help='model name')

    parser.add_argument('--input-size', type=int, default=8000,
                        help='input size in samples')

    args = parser.parse_args()
    if not os.path.exists(args.save_dir):
        os.makedirs(args.save_dir)

    model = model(batch_dim=1, input_shape=(args.input_size, 1))
    model.compile(optimizer="adam", loss="mse")
    
    # converter_model.input.set_shape((1,) + converter_model.input.shape[1:])

    converter = tf.lite.TFLiteConverter.from_keras_model(model)

    converter.optimizations = [tf.lite.Optimize.DEFAULT]
    np.random.seed(20)

    def representative_dataset_gen():
        for _ in range(100):
            # Get sample input data as a numpy array in a method of your choosing.
            yield [np.array(np.random.randn(*model.input.shape), dtype=np.float32)]

    converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS]
    converter.inference_input_type = tf.uint8
    converter.inference_output_type = tf.uint8

    converter.representative_dataset = representative_dataset_gen

    open(os.path.join(args.save_dir, args.model_name + "-uint8-quant.tflite"), "wb").write(converter.convert())
    print("Full Integer compilation successful")
```

**Comment by johannes-do** [comment]
I am having the same error when I save models during `on_train_begin()` or when I directly create it without training, thus only with the initialized weights. When training the model for at least one epoch and saving it again, the error does not appear anymore. A bit strange.

**Comment by johannes-do** [comment]
What fixes the above error for me is when I initialize the bias (in case the layer has `use_bias=True`) with anything else other than `"zero"`.

```python
x = Conv2D(8, kernel_size=3)(x) # ERROR: :334 bias->type != kTfLiteInt32 (INT8 != INT32)
x = Conv2D(8, kernel_size=3,  bias_initializer=tf.keras.initializers.GlorotNormal()) # works
```

Edit: The above workaround did not work for Conv2DTranspose.
